sklearn/model_selection/_classification_threshold.py

```diff
@@ -0,0 +1,1191 @@
+        scorer, `pos_label` can be passed as a keyword argument to
+        :func:`~sklearn.metrics.make_scorer`.
+
+    response_method : {"auto", "decision_function", "predict_proba"}, default="auto"
+        Methods by the classifier `base_estimator` corresponding to the
+        decision function for which we want to find a threshold. It can be:
+
+        * if `"auto"`, it will try to invoke, for each classifier,
+          `"predict_proba"` or `"decision_function"` in that order.
+        * otherwise, one of `"predict_proba"` or `"decision_function"`.
+          If the method is not implemented by the classifier, it will raise an
+          error.
+    """
+
+    _required_parameters = ["estimator"]
+    _parameter_constraints: dict = {
+        "estimator": [
+            HasMethods(["fit", "predict_proba"]),
+            HasMethods(["fit", "decision_function"]),
+        ],
+        "pos_label": [Real, str, "boolean", None],
+        "response_method": [StrOptions({"auto", "predict_proba", "decision_function"})],
+    }
+
+    def __init__(self, estimator, *, pos_label=None, response_method="auto"):
+        self.estimator = estimator
+        self.pos_label = pos_label
+        self.response_method = response_method
+
+    @_fit_context(
+        # *ThresholdClassifier*.estimator is not validated yet
+        prefer_skip_nested_validation=False
+    )
+    def fit(self, X, y, **params):
+        """Fit the classifier.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training data.
+
+        y : array-like of shape (n_samples,)
+            Target values.
+
+        **params : dict
+            Parameters to pass to the `fit` method of the underlying
+            classifier and to the `objective_metric` scorer.
+
+        Returns
+        -------
+        self : object
+            Returns an instance of self.
+        """
+        _raise_for_params(params, self, None)
+
+        X, y = indexable(X, y)
+
+        y_type = type_of_target(y, input_name="y")
+        if y_type != "binary":
+            raise ValueError(
+                f"Only binary classification is supported. Unknown label type: {y_type}"
+            )
+
+        if self.response_method == "auto":
+            self._response_method = ["predict_proba", "decision_function"]
+        else:
+            self._response_method = self.response_method
+
+        self._fit(X, y, **params)
+
+        if hasattr(self.estimator_, "n_features_in_"):
+            self.n_features_in_ = self.estimator_.n_features_in_
+        if hasattr(self.estimator_, "feature_names_in_"):
+            self.feature_names_in_ = self.estimator_.feature_names_in_
+
+        return self
+
+    @property
+    def classes_(self):
+        """Classes labels."""
+        return self.estimator_.classes_
+
+    @available_if(_estimator_has("predict_proba"))
+    def predict_proba(self, X):
```

Comment:
ideally all these methods should be routing to the sub-estimator's

Reply:
It looks straightforward for all those prediction from the base class. However, I think this is more tricky for the `predict` method: `predict` from the meta-estimator will delegate to the `response_method` of `estimator` meaning that we need to route from `predict` to `predict_proba`/`decision_function` that should be fine. But we need to modify `_get_response_values` to accept metadata to be able to pass those info if I'm not mistaken.

Could we instead only support metadata routing for `fit` for the moment and I'll make subsequent PRs to modify `_get_response_values` and then add support to meta-estimator that use this function?

Reply:
Happy to have it in a separate PR.